Mypy reports error "Cannot assign to a method" when using tkinter report_callback_exception

Hello,

When mypy sees tkinter exception handling like this
`root.report_callback_exception = mainloop_exc_handler`
where mainloop_exc_handler is a function or method to handle exceptions under tkinter mainloop, it reports the error "Cannot assign to a method".

I'm using mypy 0.521 with Python 3.5.4 32b on a Windows 7 SP1 x64.

I was told to open the issue here from the mypy github.

Best regards,

JM

@srittau Can we reopen this? Currently I can't get this to work. Here's a minimal example:

```python3
import tkinter
from types import TracebackType
from typing import Type

def error_handler(typ: Type[BaseException], val: BaseException, tb: TracebackType):
    pass

root = tkinter.Tk()
root.report_callback_exception = error_handler
```

```
$ mypy --custom-typeshed-dir=. asdf.py
asdf.py:11: error: Cannot assign to a method
asdf.py:11: error: Invalid self argument "Tk" to attribute function "report_callback_exception" with type "Callable[[Type[BaseException], BaseException, TracebackType], Any]"
asdf.py:11: error: Incompatible types in assignment (expression has type "Callable[[Type[BaseException], BaseException, TracebackType], Any]", variable has type "Callable[[BaseException, TracebackType], Any]")
Found 3 errors in 1 file (checked 1 source file)
```

--- a/stdlib/3/tkinter/__init__.pyi
+++ b/stdlib/3/tkinter/__init__.pyi
@@ -581,7 +581,8 @@ class Tk(Misc, Wm):
     def loadtk(self) -> None: ...  # differs from _tkinter.TkappType.loadtk
     def destroy(self) -> None: ...
     def readprofile(self, baseName: str, className: str) -> None: ...
-    report_callback_exception: Callable[[Type[BaseException], BaseException, TracebackType], Any]
+    # This really shouldn't be Optional. It's a hack to make assigning to Tk().report_callback_exception work.
+    report_callback_exception: Optional[Callable[[Type[BaseException], BaseException, TracebackType], Any]]
     # Tk has __getattr__ so that tk_instance.foo falls back to tk_instance.tk.foo
     # Please keep in sync with _tkinter.TkappType
     call: Callable[..., Any]

--- a/stdlib/3/tkinter/__init__.pyi
+++ b/stdlib/3/tkinter/__init__.pyi
@@ -4,7 +4,7 @@ from enum import Enum
 from tkinter.constants import *  # comment this out to find undefined identifier names with flake8
 from tkinter.font import _FontDescription
 from types import TracebackType
-from typing import Any, Callable, Dict, Generic, List, Optional, Tuple, Type, TypeVar, Union, overload
+from typing import Any, Callable, Dict, Generic, List, Optional, Protocol, Tuple, Type, TypeVar, Union, overload
 from typing_extensions import Literal, TypedDict
 
 # Using anything from tkinter.font in this file means that 'import tkinter'
@@ -540,6 +540,9 @@ _TkOptionName = Literal[
     "width",
 ]
 
+class _ExceptionReportingCallback(Protocol):
+    def __call__(self, __exc: Type[BaseException], __val: BaseException, __tb: TracebackType) -> Any: ...
+
 class Tk(Misc, Wm):
     master: None
     children: Dict[str, Any]
@@ -581,8 +584,7 @@ class Tk(Misc, Wm):
     def loadtk(self) -> None: ...  # differs from _tkinter.TkappType.loadtk
     def destroy(self) -> None: ...
     def readprofile(self, baseName: str, className: str) -> None: ...
-    # This really shouldn't be Optional. It's a hack to make assigning to Tk().report_callback_exception work.
-    report_callback_exception: Optional[Callable[[Type[BaseException], BaseException, TracebackType], Any]]
+    report_callback_exception: _ExceptionReportingCallback
     # Tk has __getattr__ so that tk_instance.foo falls back to tk_instance.tk.foo
     # Please keep in sync with _tkinter.TkappType
     call: Callable[..., Any]

